illumos-utils/src/zfs.rs

```diff
@@ -493,6 +511,204 @@ pub struct DatasetEnsureArgs<'a> {
+    {
+        return Ok(());
+    }
+
+    if immutablity.can_set_immutable() {
+        make_directory_mutable(mountpoint)?;
+    }
+    ensure_mountpoint_empty(mountpoint)?;
+    if immutablity.can_set_immutable() {
+        make_directory_immutable(mountpoint)?;
+    }
+
+    // This concurrent error case is a bit exceptional: we briefly made the
+    // directory mutable, tried to empty it out, and made it immutable again.
+    // However, while this was happening, someone must have added entries to the
+    // directory.
+    //
+    // This is probably a bug on the side of "whoever is adding these files".
+    if !is_directory_empty(mountpoint)? {
+        return Err(EnsureDatasetErrorRaw::DirectoryNotEmpty(
+            mountpoint.to_path_buf(),
+        ));
+    }
+    return Ok(());
+}
+
+fn is_directory_empty(path: &Utf8Path) -> Result<bool, EnsureDatasetErrorRaw> {
+    Ok(path
+        .read_dir_utf8()
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?
+        .next()
+        .is_none())
+}
+
+fn ensure_mountpoint_empty(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    if is_directory_empty(path)? {
+        return Ok(());
+    }
+
+    let (Some(parent), Some(file)) = (path.parent(), path.file_name()) else {
+        return Err(EnsureDatasetErrorRaw::BadMountpoint(path.to_path_buf()));
+    };
+
+    // The directory is not empty. Let's make a new directory,
+    // with the "old-under-mountpoint-" prefix, and move all data there.
+
+    let prefix = format!("old-under-mountpoint-{file}");
+    let destination_dir = Utf8TempDir::with_prefix_in(prefix, parent)
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?
+        .into_path();
+
+    let entries = path
+        .read_dir_utf8()
+        .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+    for entry in entries {
+        let entry = entry
+            .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+
+        // This would not work for renaming recursively, but we're only renaming
+        // a single directory's worth of files.
+        let src = entry.path();
+        let dst = destination_dir.as_path().join(entry.file_name());
+
+        std::fs::rename(src, dst)
+            .map_err(|err| EnsureDatasetErrorRaw::CreateMountpoint(err))?;
+    }
+
+    Ok(())
+}
+
+fn make_directory_immutable(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    let mut command = std::process::Command::new(PFEXEC);
+    let cmd = command.args(&["chmod", "S+ci", path.as_str()]);
+    execute(cmd).map_err(|err| EnsureDatasetErrorRaw::MakeImmutable(err))?;
+    Ok(())
+}
+
+fn make_directory_mutable(
+    path: &Utf8Path,
+) -> Result<(), EnsureDatasetErrorRaw> {
+    let mut command = std::process::Command::new(PFEXEC);
+    let cmd = command.args(&["chmod", "S-ci", path.as_str()]);
+    execute(cmd).map_err(|err| EnsureDatasetErrorRaw::MakeImmutable(err))?;
```

Comment:
Similar error variant nit as above - maybe this should be `MakeMutable` to distinguish it from failures in `make_directory_immutable()`?